sdk/python/kfp/compiler/pipeline_spec_builder.py

```diff
@@ -1626,8 +1629,98 @@ def write_pipeline_spec_to_file(pipeline_spec: pipeline_spec_pb2.PipelineSpec,
 
     elif package_path.endswith(('.yaml', '.yml')):
         with open(package_path, 'w') as yaml_file:
+            yaml_file.write(yaml_comments)
             yaml.dump(json_dict, yaml_file, sort_keys=True)
 
     else:
         raise ValueError(
             f'The output path {package_path} should end with ".yaml".')
+
+
+def extract_comments_from_pipeline_spec(pipeline_spec: dict,
+                                        pipeline_description: str) -> str:
+    map_parameter_types = {
+        'NUMBER_INTEGER': int.__name__,
+        'NUMBER_DOUBLE': float.__name__,
+        'STRING': str.__name__,
+        'BOOLEAN': bool.__name__,
+        'LIST': list.__name__,
+        'STRUCT': dict.__name__
+    }
+
+    map_headings = {
+        'inputDefinitions': '# Inputs:',
+        'outputDefinitions': '# Outputs:'
+    }
+
+    def collect_pipeline_signatures(root_dict: dict,
+                                    signature_type: str) -> List[str]:
+        comment_strings = []
+        if signature_type in root_dict:
+            signature = root_dict[signature_type]
+            comment_strings.append(map_headings[signature_type])
+
+            # Collect data
+            array_of_signatures = []
+            for parameter in signature.get('parameters', {}):
```

Comment:
Separately, this function could be simplified with `for parameter_name, parameter_body in signature['parameters'].items():`, preventing the need to reindex into three times thereafter with `signature['parameters'][parameter]`.

This also enables a clean merge of declaration of and assignment to `data` for improved readability:

```python
data = {'name': parameter_name,
        'parameterType': map_parameter_types[parameter_body['parameterType']],
        'defaultValue': parameter_body['defaultValue'],
        }
```

Reply:
I use `.get` in the first line since some pipeline inputs/outputs may or may not have parameters, so it handles both. If it does have the parameters then we enter the for loop thats why i don't use a `.get` in the for loop anymore since i am sure `signature['parameters'] exists. Also for the second point if i were to collect the data like that, i would not be able to handle sorting the `default_value` of list and dicts which would trigger remote test failures

Reply:
Makes sense on the `.get()`. I think the `.items()` suggesting still applies.

For sorting, can you investigate where the source of non-determinism is between the local/remote tests?